docs/DifferentiableTypes.md

```diff
@@ -0,0 +1,289 @@
+    /// vector. In Riemannian geometry (mathematics), this represents an
+    /// exponential map.
+    func moved(along direction: TangentVector) -> Self
+
+    /// Convert a cotangent vector to its corresponding tangent vector.
+    func tangentVector(from cotangent: CotangentVector) -> TangentVector
+}
+```
+
+Mathematically, `Differentiable` represents a differentiable manifold: this is a technical term for smooth-surfaced objects like spheres and generalizes types that are compatible with differentiation, like `Float`, `Double`, `Tensor`, and `SIMD4<Float>`. This definition comes from differential geometry and is quite technical, and not all details are relevant for most use cases.
+
+<p align="center">
+  <img src="https://upload.wikimedia.org/wikipedia/commons/3/37/Pushforward.svg" align=center>
+  <br>
+  <sub>Image showing two differentiable manifolds: a sphere and a spheroid.</sub>
+  <br>
+  <sub>From https://en.wikipedia.org/wiki/Pushforward_(differential).</sub>
+</p>
+
+Here are the important points about the `Differentiable` protocol:
+* `associatedtype TangentVector`: represents the type of directional derivatives computed via forward-mode differentiation.
+* `associatedtype CotangentVector`: represents the type of gradient values computed via reverse-mode differentiation.
+* `CotangentVector` types are used and produced by differential operators like `gradient` and `pullback`.
+* `var allDifferentiableVariables: AllDifferentiableVariables`: represents all differentiable variables in an instance of the conforming type, where `AllDifferentiableVariables` is the type of all differentiable variables.
+  * The motivation/design behind this is related to key path iteration .
+* `TangentVector`, `CotangentVector`, and `AllDifferentiableVariables` are closely related.
+  * All three associated types must themselves conform to `Differentiable`.
+  * The `Differentiable` protocol associated types of the associated types themselves are defined to be mathematically correct.
+    * `Foo.TangentVector.TangentVector` is `Foo.TangentVector` itself.
+    * `Foo.CotangentVector.TangentVector` is `Foo.CotangentVector` itself.
+    * `Foo.TangentVector.CotangentVector` is `Foo.CotangentVector`.
+    * `Foo.CotangentVector.CotangentVector` is `Foo.TangentVector`.
+    * `Foo.AllDifferentiableVariables` has the same `TangentVector` and `CotangentVector` as `Foo`.
+  * Additionally, `TangentVector` and  `CotangentVector` must conform to `AdditiveArithmetic`, so that they can be zero-initialized and accumulated via addition. These are necessary to perform the chain rule of differentiation.
+* Manifold operations.
+  * These currently involve `func tangentVector(from cotangent: CotangentVector)` and `func moved(along cotangent: CotangentVector`). These operations can be useful for implementing manifold-related algorithms, like optimization on manifolds, but are not relevant for simple differentiation use cases.
```

Comment:
* `moved(along:)` takes a tangent, not a cotangent.
* If you write the function decl, then it would be more appropriate to write the return type as well. Otherwise, using their short names seems better.

```suggestion
  * These currently involve `tangentVector(from:)` and `moved(along:)`. These operations can be useful for implementing manifold-related algorithms, like optimization on manifolds, but are not relevant for simple differentiation use cases.
```

Reply:
Changed to abbreviated names, as suggested.